sdks/python/apache_beam/typehints/native_type_compatibility.py

```diff
@@ -176,6 +177,14 @@ def convert_to_beam_type(typ):
   Raises:
     ValueError: The type was malformed.
   """
+  # Convert `int | float` to typing.Union[int, float]
+  # pipe operator as Union and types.UnionType are introduced
+  # in Python 3.10.
+  # GH issue: https://github.com/apache/beam/issues/21972
+  if (sys.version_info.major == 3 and
+      sys.version_info.minor >= 10) and (isinstance(typ, types.UnionType)):
+    typ = typing.Union[typ]
```

Comment:
Let's add a test for this functionality. it would have to be enabled only for 3.10+.